### ISSUE TITLE
PHP support

### ISSUE THREAD

**Issue by kaloyan-raev** [eclipse-che/che#2590]
This issue collects all items to be done in order to provide additional PHP support within Che.
- [x] Basic PHP plugin #2160
- [x] Enable syntax highlighting for *.phtml files #2263
- [x] Register [HvyIndustries/crane](https://github.com/HvyIndustries/crane) as the PHP language server #2282
- [x] Replace [HvyIndustries/crane](https://github.com/HvyIndustries/crane) with [felixfbecker/php-language-server](https://github.com/felixfbecker/php-language-server) as the PHP language server #2830
  - [x] InitializeParams should include an empty ClientCapabilities #2314
- [x] Enable PHP and JSON language server agents for all PHP stacks #2639
- [x] Add Zend stack (PHP 7, Zend Server and Z-Ray) #2700
- [ ] Add PHP + database compose stack with example
  Blocked by:
  - [x] Fix for using the official mysql image in composite stacks #3049
  - [ ] DB data is lost after workspace restart #3054
- [x] Composer project importer #2785
- [ ] Add more project templates:
  - [x] [Zend Framework](https://github.com/zendframework/ZendSkeletonApplication) #2785
  - [ ] [Apigility](https://github.com/zfcampus/zf-apigility-skeleton)
  - [x] [Zend Expressive](https://github.com/zendframework/zend-expressive-skeleton) #2785
  - [x] [Astro Splash demo app](https://github.com/AndrewCarterUK/AstroSplash) #2700
- [x] Out-of-the-box code completion and validation in the `composer.json` editor #2474
- [x] Integrate Zend Debugger https://github.com/eclipse/che/pull/3202
  - [x] Install Zend Debugger in the default PHP stack eclipse/che-dockerfiles#55
  - [x] Add "Run/Debug PHP script" commands to the PHP and Zend stacks #3277
  - [x] Documentation for PHP Debugging codenvy/che-docs#37
- [ ] Integrate Xdebug
- [x] Unit testing with PHPUnit #3987
  - [x] Decouple testing framework API from Java test specific frameworks #3452
  - [x] Install PHPUnit runtime in the PHP stack https://github.com/eclipse/che-dockerfiles/pull/62
- [x] Add Zend to the contributors list on the marketing home page.
- [x] Update PHP getting started guide in docs to cover all PHP scenarios

General bugs affecting the PHP support too:
- [x] Editor does not filter code completion proposals from language server #2592
- [x] Formatting the whole document makes a big mess #2437
- [x] Language Server formatting options are hardcoded #2728
- [x] Performance issues with code completion #2815
- [x] Render hover messages according to LSP specification #2856
- [x] Add commands to workspace when importing project samples #2976 
- [x] Issues with documentation popup for LSP code completion #3218

**Comment by kaloyan-raev** [comment]
@TylerJewell I started collecting the list of items related to the PHP support. I hope you and the team find it useful.

**Comment by TylerJewell** [comment]
There are some other marketing and documentation issues that will be needed.  I added them to the list for consideration.

**Comment by TylerJewell** [comment]
@kaloyan-raev - if we plan to make a 5.0 announcement release for CheConf, can we think about which issues are likely to be ready and which ones are not likely to be ready in time for the event?  It will affect how we message PHP support in the keynote that I give.  Also, it would be nice for us to launch a PHP solution page - similar to what we did for SAP web IDE or Codenvy on eclipse.org/che.  If you can create the HTML page + content, we'll get our marketing team to get it integrated into the site.

**Comment by kaloyan-raev** [comment]
@TylerJewell All "checked" items are complete - merged in master.

@bartlomiej-laczkowski is working on the Zend Debugger integration. We already have breakpoints working, stepping and showing variables state. I hope we can have the PR in a week.

@mniewrzal is helping with [felixfbecker/php-language-server](https://github.com/felixfbecker/php-language-server). There is also a great community effort there. The language server improves rapidly. I believe we can have it as a replacement of the Crane server within a week.

These are the two major improvements for the moment. When is the deadline for new major PRs for the 5.0 release?

I'll take a look at the existing documentation. Perhaps, I will open another issue to discuss the changes.

**Comment by TylerJewell** [comment]
You probably need to have a PR open by 10/31 for it to have a chance to be merged for the milestone build that we'll use for promotion purposes at CheConf.  We'll try to issue that release, whether it's M7 or M8 by November 8th.  We are not sure if that will be the GA build or not, but M5 is pretty stable and M6 is focused on exclusively customer issues, so things are rapidly moving in that direction.

**Comment by kaloyan-raev** [comment]
OK. 10/31 sounds good to me.

**Comment by slemeur** [comment]
@kaloyan-raev Are you agree to close this epic as most of the items have been covered? Maybe another epic should be opened to highlight potential advanced capabilities that you would like to introduce.

**Comment by kaloyan-raev** [comment]
@slemeur Yes. Let's close it.

**Comment by madalinignisca** [comment]
Will Xdebug be ever supported?

Zend Debugger won't run with PHP 7.2 and it's not nice to have to wait for Zend to work it out on each new PHP release to update their closed source extension. From their history it can be many months...

By the way, is there any reasons not to make Xdebug a priority and default as it's 100% open source and support the closed source Zend Debugger extension?
